vumi_message_store/batch_info_cache.py

```diff
@@ -352,31 +352,117 @@ def rebuild_cache(self, batch_id, qms, page_size=None):
         """
         Rebuild the cache using the provided IQueryMessageStore implementation.
         """
-        # TODO: Make this less naive.
         yield self.clear_batch(batch_id)
         yield self.batch_start(batch_id)
 
+        yield self._rebuild_inbound_messages(batch_id, qms, page_size)
+        yield self._rebuild_outbound_messages(batch_id, qms, page_size)
+        yield self._rebuild_events(batch_id, qms, page_size)
+
+    @Manager.calls_manager
+    def _rebuild_inbound_messages(self, batch_id, qms, page_size=None):
+        """
+        Rebuild the cache by loading the latest inbound messages for the given
+        batch into the cache and counting all the messages.
+        """
         inbound_page = yield qms.list_batch_inbound_keys_with_addresses(
             batch_id, max_results=page_size)
+        count = 0
+        recents_added = False
+        from_addrs = set()
         while inbound_page is not None:
             for key, timestamp, from_addr in inbound_page:
-                yield self.add_inbound_message_key(
-                    batch_id, key, to_timestamp(timestamp))
-                yield self.add_from_addr(batch_id, from_addr)
+                # Treat the most recent messages as though we were recording
+                # them in flight.
+                if not recents_added:
+                    yield self.add_inbound_message_key(
+                        batch_id, key, to_timestamp(timestamp))
+                    yield self.add_from_addr(batch_id, from_addr)
+                    if count == self.TRUNCATE_MESSAGE_KEY_ZSET_AT:
+                        recents_added = True
+                        count = 0
+                else:
+                    from_addrs.add(from_addr)
+                count += 1
```

Comment:
Should this increment not happen at the start of the loop instead? At the moment, it seems we're incrementing the count for the last "recent" message after zeroing the account.